doc/src/guides/custom-functions.md

```diff
@@ -0,0 +1,1773 @@
+can deduce automatically given its [known
+facts](assumptions-guide-predicates).
+
+```py
+>>> versin(1).is_real
+True
+```
+
+The assumptions system is often able to deduce more than you might think.
+For example, from the above, it can deduce that `versin(2*n*pi)` is zero when
+`n` is an integer.
+
+```py
+>>> n = symbols('n', integer=True)
+>>> versin(2*n*pi).is_zero
+True
+```
+
+It's always worth checking if the assumptions system can deduce something
+automatically before manually coding it.
+
+Finally, a word of warning: be very careful about correctness when coding
+assumptions. Make sure to use the exact
+[definitions](assumptions-guide-predicates) of the various assumptions, and
+always check that you're handling `None` cases correctly with the fuzzy
+three-valued logic functions. Incorrect or inconsistent assumptions can lead
+to subtle bugs. It's recommended to use unit tests to check all the various
+cases whenever your function has a nontrivial assumption handler. All
+functions defined in SymPy itself are required to be extensively tested.
+
+(custom-functions-evalf)=
+### Numerical Evaluation with `evalf()`
+
+<!-- TODO: this goes over the basics, but it might be useful to have a
+     separate guide dedicated to numerical evaluation. -->
+
+Here we show how to define how a function should numerically evaluate to a
+floating point {class}`~.Float` value, for instance, via `evalf()`.
+Implementing numerical evaluation enables several behaviors in SymPy. For
+example, once `evalf()` is defined, you can plot your function, and things
+like inequalities can evaluate to explicit values.
+
+If your function has the same name as a function in
+[mpmath](https://mpmath.org/doc/current/), which is the case for most
+functions included with SymPy, numerical evaluation will happen automatically
+and you do not need to do anything.
+
+If this is not the case, numerical evaluation can be specified by defining the
+method `_eval_evalf(self, prec)`, where `prec` binary precision of the input.
+The method should return the expression evaluated to the given precision, or
+`None` if this is not possible.
+
+```{note}
+
+The `prec` argument to `_eval_evalf()` is the *binary* precision, that is, the
+number of bits in the floating-point representation. This differs from the
+first argument to the `evalf()` method, which is the *decimal* precision, or
+`dps`. For example, the default binary precision of `Float` is 53,
+corresponding to a decimal precision of 15. Therefore, if your `_eval_evalf()`
+method recursively calls evalf on another expression, it should call
+`expr._eval_evalf(prec)` rather than `expr.evalf(prec)`, as the latter will
+incorrectly use `prec` as the decimal precision.
+
+```
+
+We can define numerical evaluation for [our example $\operatorname{versin}(x)$
+function](custom-functions-versine-definition) by recursively evaluating
+$2\sin^2\left(\frac{x}{2}\right)$, which is a more numerically stable way of writing $1 -
+\cos(x)$.
```

Comment:
Maybe not worth mentioning here but perhaps check out the discussion in #20479 and #20493

Reply:
See also #20321